### PR TITLE
feat: Add isDisabled to Button column and isVisible to all Table widget Columns

### LINKED ISSUE #3250: [Feature] Add isVisible property for all columns & isDisabled property for button column in Table Widget
## Summary
Add disable property on a row button which can be controlled via JS

## Motivation
For some row actions that should not enabled, we should be able to control button disability via js

### LINKED ISSUE #5943: [Feature] Table Columns should Have Column properties and Cell properties
## Summary

Properties like Visible and Disabled are applied to the column and not individual cells in the column while Text align and and other stylings are applied to cells, this should be properly separated

Screenshot 2021-07-16 at 10 38 53

## Motivation
Would cause confusion for Users

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

> Add new column properties to Table

Fixes  #3250 
Fixes #5943 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feat/isVisible-and-isdisable-property-for-table-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 54.84 **(0.01)** | 36.63 **(-0.04)** | 33.27 **(0.02)** | 55.42 **(0.01)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx | 67.82 **(-0.78)** | 21.85 **(-0.16)** | 50 **(-0.91)** | 67.82 **(-0.78)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableUtilities.tsx | 44.17 **(-0.75)** | 22.89 **(-1.16)** | 18.75 **(0)** | 42.73 **(-0.79)**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx | 47.95 **(1.28)** | 29.63 **(0)** | 7.69 **(0)** | 47.89 **(1.31)**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/JSONtoForm.tsx | 16.2 **(0.72)** | 20.29 **(2.11)** | 4.35 **(0.5)** | 17.32 **(0.89)**
 :red_circle: | app/client/src/utils/WidgetPropsUtils.tsx | 72.18 **(-0.56)** | 64.52 **(-0.48)** | 67.03 **(1.12)** | 71.15 **(-0.28)**
 :red_circle: | app/client/src/widgets/TableWidget/index.tsx | 27.41 **(-0.38)** | 6.29 **(-0.71)** | 25.49 **(-0.51)** | 27.58 **(-0.42)**</details>

### CODE REVIEW

File: app/client/src/constants/WidgetConstants.tsx [hunk 1]
```diff
@@ -96,7 +96,7 @@ export const layoutConfigurations: LayoutConfigurations = {
   FLUID: { minWidth: -1, maxWidth: -1 },
 };
 
-export const LATEST_PAGE_VERSION = 23;
+export const LATEST_PAGE_VERSION = 24;
```
Comment (hetunandu) [inline app/client/src/constants/WidgetConstants.tsx:99]: We would need to update the tests again right @marks0351?

Comment (marks0351) [reply app/client/src/constants/WidgetConstants.tsx:99]: thanks for reminding, that's true. @Tooluloope make sure to update the dsl versions in https://github.com/appsmithorg/appsmith/blob/release/app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/ForkApplication_spec.js#L8
and 
https://github.com/appsmithorg/appsmith/blob/release/app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js#L9

Comment (hetunandu) [reply app/client/src/constants/WidgetConstants.tsx:99]: @marks0351 When will we make the change we discussed to avoid such problems?

Comment (marks0351) [reply app/client/src/constants/WidgetConstants.tsx:99]: @NandanAnantharamu u r working on this already right?

Comment (NandanAnantharamu) [reply app/client/src/constants/WidgetConstants.tsx:99]: @marks0351 facing some network issue will update this spec asap once my network is stable

Comment (Tooluloope) [reply app/client/src/constants/WidgetConstants.tsx:99]: Any update @NandanAnantharamu , I’ve update the version also.

**Comment by Tooluloope** [comment]
Re-running test cause it's failing

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Review by somangshu — DISMISSED** [review]

**Comment by somangshu** [comment]
@Tooluloope please merge the latest release and resolve the above conflicts, Let me know if any help is required!

**Review by somangshu — APPROVED** [review]

**Comment by mohanarpit** [comment]
@Tooluloope Please merge the latest release into this branch and try to close this PR. The flaky tests should have been fixed now.

**Comment by somangshu** [comment]
@Tooluloope need to resolve the conflict and.

**Review by somangshu — APPROVED** [review]

File: app/client/cypress/manual_TestSuite/GoogleSheet_API_Spec.js [hunk 2]
```diff
@@ -3,37 +3,37 @@ const queryEditor = require("../../../locators/QueryEditor.json");
 let datasourceName;
 
 describe("Test Ideas for GooglSheet API", function() {
-    it("Add a Datasource", function() {
```
Comment (somangshu) [inline app/client/cypress/manual_TestSuite/GoogleSheet_API_Spec.js:6]: hey @Tooluloope if i am not wrong, Was the conflict because of the indentation.

Comment (Tooluloope) [reply app/client/cypress/manual_TestSuite/GoogleSheet_API_Spec.js:6]: No, it wasn't.
Nope, here was the Conflict.

Screenshot 2021-06-29 at 15 43 47

**Comment by somangshu** [comment]
@marks0351 kindly review

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js [hunk 3]
```diff
@@ -8,6 +8,7 @@ let parentApplicationDsl;
 
 describe("Duplicate application", function() {
   before(() => {
+    dsl.dsl.version = 27; // latest migrated version
```
Comment (marks0351) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js:11]: i dont think this is needed now. pls remove and test it out. same goes for ForkApplication_spec.js

**Review by marks0351 — APPROVED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js (hunk 3 above)
Comment (somangshu) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js:11]: @NandanAnantharamu @yatinappsmith can this is handled separately?

Comment (NandanAnantharamu) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js:11]: i think @marks0351 has rightly said he has fixed this issue this is not needed anymore for both fork / duplicate

Comment (somangshu) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js:11]: My question is is we need to remove it in this PR or do it seperately?

Comment (marks0351) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Applications/DuplicateApplication_spec.js:11]: we need to remove them @somangshu

**Review by marks0351 — CHANGES_REQUESTED** [review]

**Review by somangshu — DISMISSED** [review]
Running tests

**Review by marks0351 — DISMISSED** [review]

**Comment by vivekverma2312** [comment]
@Tooluloope Observations on deploy preview:
1. Enabling `Disable` property in column property pane do not make the button grey in colour
2. Binding with `currentRow` is throwing error for visible and disable property

https://www.loom.com/share/35e877a3f66944139e929ba2b3c58937

**Comment by vivekverma2312** [comment]
@Tooluloope Above issues still exist.

**Comment by Tooluloope** [comment]
> @Tooluloope Above issues still exist.

@vivekverma2312, the Button issue has been fixed.
As for the currentRow binding, I think it's a general Table issue.

CurrentRow binding shows undefined for all other fields except computed value

File: app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx [hunk 4]
```diff
@@ -280,6 +280,10 @@ export const ActionWrapper = styled.div<{
     .bp3-button span {
       font-weight: 400;
     }
+    &&& .bp3-disabled {
+      color: #768896;
```
Comment (somangshu) [inline app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx:284]: Can we get these from the default theme?

Comment (Tooluloope) [reply app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx]: Fixed

**Comment by vivekverma2312** [comment]
> > @Tooluloope Above issues still exist.
> 
> @vivekverma2312, the Button issue has been fixed.
> As for the currentRow binding, I think it's a general Table issue.
> 
> CurrentRow binding shows undefined for all other fields except computed value

@Tooluloope Button issue is working fine.
Checked currentRow for Button and label colour property and it worked.
Screenshot 2021-07-13 at 6 21 56 PM

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@Tooluloope please resolve the conflicts here

**Comment by somangshu** [comment]
@marks0351 need you review here!

**Comment by marks0351** [comment]
@Tooluloope can u fix the jest test that are failing?

**Review by marks0351 — DISMISSED** [review]

**Comment by somangshu** [comment]
@Tooluloope related test failing here

**Comment by Tooluloope** [comment]
> @Tooluloope related test failing here

Fixed @somangshu

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Review by marks0351 — DISMISSED** [review]

**Comment by somangshu** [comment]
Quoting here for [ref](https://github.com/appsmithorg/appsmith/issues/5943#issuecomment-884703527)

**Review by somangshu — APPROVED** [review]
approving for tests to run

File: app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts [hunk 5]
```diff
@@ -151,7 +151,6 @@ export interface ColumnProperties {
   onOptionChange?: string;
   displayText?: string;
   isCellVisible?: boolean;
-  isColumnVisible?: boolean;
```
Comment (Tooluloope) [inline app/client/src/components/designSystems/appsmith/TableComponent/Constants.ts]: @marks0351 Some updates were made to the issue,
See Ref here https://github.com/appsmithorg/appsmith/issues/5943

File: app/client/src/entities/Widget/utils.test.ts [hunk 6]
```diff
@@ -153,8 +161,14 @@ describe("getAllPathsFromPropertyConfig", () => {
           EvaluationSubstitutionType.TEMPLATE,
         "primaryColumns.status.buttonStyle":
           EvaluationSubstitutionType.TEMPLATE,
+        // "primaryColumns.status.isVisible": EvaluationSubstitutionType.TEMPLATE,
```
Comment (marks0351) [inline app/client/src/entities/Widget/utils.test.ts:164]: can u pls remove comments if they are unnecessary.

**Review by marks0351 — APPROVED** [review]

**Comment by somangshu** [comment]
Re-running all tests once.

@Tooluloope there is a migration required here, The old table shows visible as false by default and the column remains visible, This will need to be fixed.

**Review by somangshu — APPROVED** [review]
LGTM

Deploy preview tested

**Comment by areyabhishek** [comment]
@Tooluloope @vivekverma2312 The eye icon is off-center and it's smaller in size that it is supposed to be:
Screen Shot 2021-09-05 at 1 21 47 PM

Link to Figma file of the table property pane: https://www.figma.com/file/1G9ShIQgRloGvcS6kLgw2O/Table-widget-property-pane?node-id=1%3A3

**Comment by Tooluloope** [comment]
> @Tooluloope @vivekverma2312 The eye icon is off-center and it's smaller in size that it is supposed to be:
> Screen Shot 2021-09-05 at 1 21 47 PM
> 
> Link to Figma file of the table property pane: https://www.figma.com/file/1G9ShIQgRloGvcS6kLgw2O/Table-widget-property-pane?node-id=1%3A3

@areyabhishek Currently not on my PC, could you create an issue for it and assign me.

**Pull request merged**